File: app/src/main/java/blue/koenig/kingsfinances/view/ChartHelper.java
package blue.koenig.kingsfinances.view;

import android.graphics.Color;

import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet;
import com.koenig.commonModel.User;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import blue.koenig.kingsfinances.model.calculation.StatisticEntry;

/**
 * Created by Thomas on 05.01.2018.
 */

public class ChartHelper {

    public static List<String> entrysToMonthXValues(List<StatisticEntry> statisticEntryList) {
        ArrayList<String> xEntrys = new ArrayList<>(statisticEntryList.size());
        for (StatisticEntry debt : statisticEntryList) {
            String dateString = debt.getDate().toString("MM/yy");
            xEntrys.add(dateString);
        }

        return xEntrys;

    }

    public static LineData entrysToLineData(List<StatisticEntry> debts, String userId, int color) {
        // One debts not possible
        if (debts.size() <= 1) return new LineData();
        Map<User, Integer> debtsMap = debts.get(1).getEntryMap();
        List<Integer> colors = new ArrayList<>(debts.size());
        ArrayList<ILineDataSet> dataSets = new ArrayList<>();
        for (User user : debtsMap.keySet()) {
            if (user.getId().equals(userId)) {
                ArrayList<Entry> userEntrys = new ArrayList<>(debts.size());
                int i = 0;
                for (StatisticEntry debt : debts) {
                    Integer entryForNullabe = debt.getEntryForNullabe(user);
                    if (entryForNullabe != null) {
                        int value = entryForNullabe;
                        userEntrys.add(new Entry(i, value / 100f));
                        if (value < 0) {
                            colors.add(Color.RED);
                        } else colors.add(Color.GREEN);
                    }

                    i++;
                }
                LineDataSet barDataSet = new LineDataSet(userEntrys, user.getName());
                //barDataSet.setColors(colors);
                barDataSet.setColor(color);
                //barDataSet.setBarBorderWidth(20);

                barDataSet.setValueTextColor(Color.WHITE);
                dataSets.add(barDataSet);
            }
        }

        LineData lineData = new LineData(dataSets);
        //lineData.setBarWidth(0.04f);

        return lineData;
    }

    public static LineData entrysToLineData(List<StatisticEntry> entries, int[] colors) {
        // One entries not possible
        if (entries.size() <= 1) return new LineData();
        Map<User, Integer> debtsMap = entries.get(1).getEntryMap();
        ArrayList<ILineDataSet> dataSets = new ArrayList<>();

        int userCount = 0;
        for (User user : debtsMap.keySet()) {
            int i = 0;
            ArrayList<Entry> userEntrys = new ArrayList<>(entries.size());
            for (StatisticEntry debt : entries) {
                Integer entryForNullabe = debt.getEntryForNullabe(user);
                if (entryForNullabe != null) {
                    int value = entryForNullabe;
                    userEntrys.add(new Entry(i, value / 100f));
                }
                i++;
            }
            LineDataSet barDataSet = new LineDataSet(userEntrys, user.getName());
            barDataSet.setColor(colors[userCount % colors.length]);
            barDataSet.setValueTextColor(Color.WHITE);
            dataSets.add(barDataSet);
            userCount++;
        }

        LineData lineData = new LineData(dataSets);
        return lineData;
    }
}

File: app/src/main/java/blue/koenig/kingsfinances/view/ChartHelper.kt
package blue.koenig.kingsfinances.view

import android.graphics.Color
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet
import com.koenig.commonModel.User
import com.koenig.commonModel.finance.statistics.MonthStatistic
import org.joda.time.YearMonth
import java.util.*

/**
 * Created by Thomas on 05.01.2018.
 */

object ChartHelper {

    fun entrysToMonthXValues(statisticEntryList: List<YearMonth>): List<String> {
        val xEntrys = ArrayList<String>(statisticEntryList.size)
        for (debt in statisticEntryList) {
            val dateString = debt.toString("MM/yy")
            xEntrys.add(dateString)
        }

        return xEntrys

    }


    fun mapToLineData(entries: List<MonthStatistic>, colors: IntArray, users: List<User>): LineData {
        // One entries not possible
        if (entries.size <= 1) return LineData()
        val dataSets = ArrayList<ILineDataSet>()

        for ((userCount, user) in users.withIndex()) {
            val userEntrys = ArrayList<Entry>(entries.size)
            entries.forEachIndexed { index, monthStatistic ->
                monthStatistic.entryMap[user]?.let {
                    userEntrys.add(Entry(index.toFloat(), it / 100f))
                }
            }

            val barDataSet = LineDataSet(userEntrys, user.name)
            barDataSet.color = colors[userCount % colors.size]
            barDataSet.valueTextColor = Color.WHITE
            if (barDataSet.entryCount > 0) dataSets.add(barDataSet)
        }

        return LineData(dataSets)
    }
}
